src/Symfony/Component/PropertyInfo/Util/LegacyTypeConverter.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\PropertyInfo\Util;
+
+use Symfony\Component\PropertyInfo\Type as LegacyType;
+use Symfony\Component\TypeInfo\Type;
+
+/**
+ * @internal
+ */
+class LegacyTypeConverter
+{
+    /**
+     * @param LegacyType[]|null $legacyTypes
+     */
+    public static function toTypeInfoType(?array $legacyTypes): ?Type
+    {
+        if (null === $legacyTypes || [] === $legacyTypes) {
+            return null;
+        }
+
+        $nullable = false;
+        $types = [];
+
+        foreach ($legacyTypes as $legacyType) {
+            if ($legacyType->isCollection() && !\in_array($legacyType->getBuiltinType(), [LegacyType::BUILTIN_TYPE_ARRAY, LegacyType::BUILTIN_TYPE_ITERABLE], true)) {
+                $typeInfoType = Type::collection(Type::object(\Iterator::class), self::toTypeInfoType($legacyType->getCollectionValueTypes()), self::toTypeInfoType($legacyType->getCollectionKeyTypes()));
```

Comment:
Using `Type::object(\Iterator::class)` looks wrong to me. It should be using either `Type::array()` or `Type::iterable()` depending on the builtin type.

Reply:
This case here is about collections where the built-in type is neither iterable nor array. But that probably means that the built-in type is an object here. I have rewritten the check accordingly.